### PR TITLE
WT-5648 add a leaf or internal page type flag to the WT_REF structure

### PR DESCRIPTION
**Pull request by keithbostic**

### CODE REVIEW

File: src/btree/bt_discard.c [hunk 1]
```diff
@@ -243,6 +243,12 @@ __wt_free_ref(WT_SESSION_IMPL *session, WT_REF *ref, int page_type, bool free_pa
     if (ref == NULL)
         return;
 
+    /*
+     * Flagging the page type in the WT_REF is a change, and the only current user of WT_REF.flags.
+     * Assert for now, to ensure we find all paths which allocate a WT_REF (02/2020).
```
Comment (agorrod) [inline src/btree/bt_discard.c]: Could you reword this to be more permanent? I don't see why we wouldn't always want this. Maybe `Ensure WT_REF structure is properly initialized - there are many places that create WT_REFs, and they must always refer to either a leaf or internal page.`

File: src/include/btmem.h [hunk 2]
```diff
@@ -820,6 +820,17 @@ struct __wt_ref {
     WT_PAGE *volatile home;        /* Reference page */
     volatile uint32_t pindex_hint; /* Reference page index hint */
 
+    uint8_t __unused[2]; /* Padding */
```
Comment (agorrod) [inline src/include/btmem.h]: I suspect you moved this to the middle of the structure so that most accesses were somewhat aligned and don't need a memory copy. It's probably worth describing that in a comment, since it's not obvious why the padding should be in the middle of the structure.

File: src/btree/bt_discard.c (hunk 1 above)
Comment (keithbostic) [reply src/btree/bt_discard.c]: Done, and the test reworked to match.

I don't see any reason to remove this unless we run out of bits  in `WT_REF.flags`, in which case we could lose one of these bits, only one is necessary.

Pushed in 9f1815a8e.

File: src/include/btmem.h (hunk 2 above)
Comment (keithbostic) [reply src/include/btmem.h]: Two reasons: as you say, we wanted the 4 `uint8_t` fields in a row, after the `uint32_t pindex_hint` field for better alignment, and second, `unused` is specifically where it is so we can naturally grow the `WT_REF.flags` to 2 bytes, to add more bits if needed.

Pushed in 57e24884f.
